packages/backend-plugin-manager/src/manager/plugin-manager.ts

```diff
@@ -0,0 +1,265 @@
+    logger: LoggerService,
+    preferAlpha: boolean = false,
+    mooduleLoader: ModuleLoader = new CommonJSModuleLoader(logger),
+  ): Promise<PluginManager> {
+    /* eslint-disable-next-line no-restricted-syntax */
+    const backstageRoot = findPaths(__dirname).targetRoot;
+    const scanner = new PluginScanner(
+      config,
+      logger,
+      backstageRoot,
+      preferAlpha,
+    );
+    const scannedPlugins = await scanner.scanRoot();
+    scanner.trackChanges();
+    const manager = new PluginManager(logger, scannedPlugins, mooduleLoader);
+
+    const dynamicPluginsPaths = scannedPlugins.map(p =>
+      fs.realpathSync(
+        path.dirname(
+          path.dirname(
+            path.resolve(url.fileURLToPath(p.location), p.manifest.main),
+          ),
+        ),
+      ),
+    );
+
+    mooduleLoader.bootstrap(backstageRoot, dynamicPluginsPaths);
+
+    scanner.subscribeToRootDirectoryChange(async () => {
+      manager._availablePackages = await scanner.scanRoot();
+      // TODO: do not store _scannedPlugins again, but instead store a diff of the changes
+    });
+    manager.plugins.push(...(await manager.loadPlugins()));
+
+    return manager;
+  }
+
+  readonly plugins: DynamicPlugin[];
+  private _availablePackages: ScannedPluginPackage[];
+
+  private constructor(
+    private readonly logger: LoggerService,
+    private packages: ScannedPluginPackage[],
+    private readonly moduleLoader: ModuleLoader,
+  ) {
+    this.plugins = [];
+    this._availablePackages = packages;
+  }
+
+  get availablePackages(): ScannedPluginPackage[] {
+    return this._availablePackages;
+  }
+
+  addBackendPlugin(plugin: BackendDynamicPlugin): void {
+    this.plugins.push(plugin);
+  }
+
+  private async loadPlugins(): Promise<DynamicPlugin[]> {
+    const loadedPlugins: DynamicPlugin[] = [];
+
+    for (const scannedPlugin of this.packages) {
+      const platform = PackageRoles.getRoleInfo(
+        scannedPlugin.manifest.backstage.role,
+      ).platform;
+
+      if (
+        platform === 'node' &&
+        scannedPlugin.manifest.backstage.role.includes('-plugin')
+      ) {
+        const plugin = await this.loadBackendPlugin(scannedPlugin);
+        if (plugin !== undefined) {
+          loadedPlugins.push(plugin);
+        }
+      } else {
+        loadedPlugins.push({
+          name: scannedPlugin.manifest.name,
+          version: scannedPlugin.manifest.version,
+          role: scannedPlugin.manifest.backstage.role,
+          platform: 'web',
+          // TODO(davidfestal): add required front-end plugin information here.
+        });
+      }
+    }
+    return loadedPlugins;
+  }
+
+  private async loadBackendPlugin(
+    plugin: ScannedPluginPackage,
+  ): Promise<BackendDynamicPlugin | undefined> {
+    const packagePath = url.fileURLToPath(
+      `${plugin.location}/${plugin.manifest.main}`,
+    );
+    try {
+      const { dynamicPluginInstaller } = await this.moduleLoader.load(
+        packagePath,
+      );
```

Comment:
Because I abstracted the loading part in  the `ModuleLoader` interface.
In the future, this will allow implementing other types of loading (secure loading, `isolated-vm`-based ?, ...).

Reply:
Eh, unlikely tbh because the loaded features are hooked in and rely on the existing JS context. Isolated runners would probably wrap the dynamic feature more directly if we need them.

Reply:
`isolated-vm` was given here as an example, but more generally this allows replacing the default loading mechanism.